### DISCUSSION TITLE
Why was GitHub releases abandoned?

### DISCUSSION THREAD

**Discussion by subnut** [Eloston/ungoogled-chromium#1279] [Q&A] [answered]
I'm curious

**Comment by wchen342** [comment]
I think it is because each platform needs to be packaged differently so it is very hard to combine them into a single release.

**Reply by subnut** [reply]
But we could just add different packages for different platforms!
If needed, we could automatically build and add the packages to the release using a GitHub Workflow!

**Comment by omentic** [comment]
It's almost certainly a combination of @wchen342 's answer and that maintaining releases is a pain, but having an official `.tar.gz` package to take and repackage in other distros would be nice (and, although builds aren't reproducible, repackages could be verified against an "official" release). The disassociation between [Ungoogled Binaries](https://ungoogled-software.github.io/ungoogled-chromium-binaries/) and the project is understandable, but also not great as you need to trust more users than just `@Eloston`.

**Reply by ilu33** [reply]
I second this and opened a feature request here https://github.com/ungoogled-software/ungoogled-chromium-portablelinux/issues/86. Building chromium is a task only major distributions can tackle.

Edit: I was not sure where to open the issue. Some discussion ended up here https://github.com/Eloston/ungoogled-chromium/issues/1368

**Reply by Eloston** [reply]
> and that maintaining releases is a pain

Pretty much this.

Nowadays we can automate this, but there isn't much value in doing so if the binaries aren't going to be attached to them.

**Reply by subnut** [reply]
@Eloston We can attach the binaries to GitHub releases, can't we? 

We can automate the building using GitHub actions... and upload the built binary to the GitHub Release.
It's all just a matter of workflows, really. (I have been doing something similar in a project of mine for 6 months)

I don't see any problem/reason to avoid the above method. Why don't we do that?

**Reply by kramred** [reply]
It's being done for the [macOS releases](https://github.com/kramred/ungoogled-chromium-macos/releases). One issue may be the long build times, the arm64 macOS binaries take 28 h to finish building; and the macOS virtual machines are more powerful than the other systems. So, for Linux and Windows it would probably also take around a day to finish building on GitHub Actions.

I do agree that this is a good way for people to trust the provided binaries more.

**Reply by ilu33** [reply]
The long build times and resources needed are exactly the reason why we can't build (and we have a lot of other packages to watch out for too). Some trusted workflow for the portable Linux build would be great, either in a approved OBS repo or on the releases page.

Edit: I was not sure where to open the issue. Some discussion ended up here https://github.com/Eloston/ungoogled-chromium/issues/1368

**Reply by omentic** [reply]
@ilu33 Taking a day to build on Github Actions seems moderately inconvenient but not a deal-breaker if it's only updated once or twice a week. Something with OBS seems even better, but can it be automatically hooked up to releases on Github?

**Reply by wchen342** [reply]
There are already several OBS repositories for popular Linux distributions ([Arch Linux](https://build.opensuse.org/package/show/home:justkidding:arch/ungoogled-chromium), [Debian/Ubuntu](https://build.opensuse.org/project/show/home:ungoogled_chromium), [Fedora/CentOS](https://build.opensuse.org/project/show/home:wchen342:ungoogled-chromium-fedora)), but they are not listed in the obvious places so they are a little hard to find for first time visitors.

@Eloston I think the pages at [https://ungoogled-software.github.io/downloads/](https://ungoogled-software.github.io/downloads/) and [https://ungoogled-software.github.io/ungoogled-chromium-binaries/](https://ungoogled-software.github.io/ungoogled-chromium-binaries/) needs to be updated to point to distribution specific repository pages.

Also maybe it is better to put distribution specific downloads in front of the link to `ungoogled-chromium-binaries` in the [`#Downloads`](https://github.com/Eloston/ungoogled-chromium#downloads) section. From my observation visitors will try to find and notice the binary download section first, and because the section says _`These binaries are provided by anyone who are willing to build and submit them`_ at the beginning they will assume there are no credible builds, despite there are OBS repos existing.

**Reply by ilu33** [reply]
I'm specifically NOT asking about distribution specific builds but about the portable version. I know about the Debian OBS but we cannot use it because it conflicts with the package provided by Debian. 
And yes, that sentence is off-putting but it's the truth, isn't it? For redistribution we really need something more trustworthy.
If github actions works, fine - might even be better than OBS, because the project's endorsement is stronger.

**Reply by wchen342** [reply]
> I'm specifically NOT asking about distribution specific builds but about the portable version.

I think you need to ask @braewoods specifically for this. AFAIK the portable version is based on Debian and he is the only one knows how to build Debian packages on OBS. @jstkdng only do Arch builds and I only do Fedora builds so we cannot help.

> If github actions works, fine - might even be better than OBS,

No, Actions are not designed for building something as big as chromium. Actions has too many arbitrary limitations which makes building chromium incredibly difficult and unnecessarily complicated. Namely the machines are limited to 4 CPUs and 16GB memory if I remember correctly, and each job has an upper limit of 6 hours. I used to try doing this in Arch repo and it is too much of a trouble (_I need to build half way, pack everything in a compressed file, and then upload it as an manifest, and spin up a new job and download the compressed file again and continue building in a new job. Due to low CPU counts this needs to be done 3-4 times for one build and it takes 12+ hours_).  This is why eventually Arch repo abandoned this method.

**Reply by subnut** [reply]
@wchen342 Oh, I see.... 
This (ie. not building in GitHub Actions due to the CPU and build time restrictions) is exactly what I wanted to know when I asked this question. Thanks!

**Reply by ilu33** [reply]
@wchen342 I know there is a Debian repository - but I was not asking for a debian package. And @braewoods presently doesn't build the portable version. 

Actually, https://ungoogled-software.github.io/ungoogled-chromium-binaries/releases/linux_portable/64bit/ is exactly what we need - but who is LordTwix? S/he does not belong to the project, has no reputation of any kind as far as I can see and the projects disclaimer on the downloads page is correct but scary. A distro cannot responsibly use that package. 

One single official build, that is generic and portable, would make redistribution possibile. No matter what codebase the distro uses, portable fits everyone.

**Reply by wchen342** [reply]
Well then you need to wait for someone who:
 - Knows how to build the portable version
 - Knows how to build on OBS (or some other build service)
at the same time. Unfortunately I do not know such a person.

My personal view, a portable version that can run on all distributions needs to bundle every library with it otherwise you *will* run into distribution specific problems. In that case I don't see why not using AppImage or Flatpak.

**Reply by ilu33** [reply]
There is absolutely no problem with libraries because we will build the package tailored for our distro. We just can't compile.

But yeah. Waiting for someone with knowledge and OBS - and with project approval (that's the main point in this discussion and it looks as if that won't happen). 
And thanks for the clarification about github actions btw.

**Reply by ghost** [reply]
@subnut GitHub actions, though in theory usable for builds, was never really intended for our requirements. A service like OBS is. Although it is limited to Linux it does handle that job very well. It has lots of resources and very high runtime limits. Given how much chromium takes to build that is a very big deal. It also enables builds for non-x86 architectures, mostly 32 and 64 bit arm. And all of this is done in native environments so we don't have to mess with the hassles that cross-compilation imposes which tends to worsen as the dependencies grow.

Originally we were using Launchpad to build stuff but that is limited to Ubuntu only and doesn't lend itself well to automated uploads so it was ultimately better to switch to OBS which has a web API we can use for submitting builds. We originally had a long discussion about OBS here: https://github.com/ungoogled-software/ungoogled-chromium-archlinux/issues/70

**Comment by subnut** [comment, marked as answer]
I am marking this discussion as answered. See the following two links for the answers -

- https://github.com/Eloston/ungoogled-chromium/discussions/1279#discussioncomment-378545
- https://github.com/Eloston/ungoogled-chromium/discussions/1279#discussioncomment-582818

I am posting them here and marking this as the answer because I don't know how to mark comments as answers.
Thank you for answering!
